src/kdalgorithms.h

```diff
@@ -585,10 +585,69 @@ auto partitioned(Container &&container, UnaryPredicate predicate)
         std::forward<Container>(container), std::forward<UnaryPredicate>(predicate));
 }
 
+// -------------------- multi_partitioned --------------------
+#if __cplusplus >= 202002L
+namespace detail {
+    template <typename ResultContainer, typename InputContainer, typename KeyFunction>
+    constexpr bool KeyFunctionMapsContainerValueToMapKey =
+        std::is_convertible_v<detail::invoke_result_t<KeyFunction, ValueType<InputContainer>>,
+                              typename ResultContainer::key_type>;
+
+    template <typename ResultContainer, typename InputContainer>
+    constexpr bool InputValueCanConvertToResultContainerValue =
+        std::is_convertible_v<ValueType<InputContainer>,
+                              typename ResultContainer::mapped_type::value_type>;
+}
+#endif
+
+template <typename ResultContainer, typename InputContainer, typename KeyFunction>
+#if __cplusplus >= 202002L
+requires std::is_invocable_v<KeyFunction, ValueType<InputContainer>> && detail::
+    KeyFunctionMapsContainerValueToMapKey<ResultContainer, InputContainer, KeyFunction> && detail::
+        InputValueCanConvertToResultContainerValue<ResultContainer, InputContainer>
+#endif
+auto multi_partitioned(InputContainer &&container, KeyFunction keyFunction)
+{
+    ResultContainer result;
+    auto iterators = read_iterator_wrapper(std::forward<InputContainer>(container));
+    for (auto it = iterators.begin(); it != iterators.end(); ++it) {
+        result[detail::invoke(keyFunction, *it)].push_back(*it);
+    }
+    return result;
+}
+
+template <typename InputContainer, typename KeyFunction>
+#if __cplusplus >= 202002L
+requires std::is_invocable_v<KeyFunction, ValueType<InputContainer>>
+#endif
+auto multi_partitioned(InputContainer &&container, KeyFunction keyFunction)
+{
+    using KeyType = remove_cvref_t<decltype(detail::invoke(keyFunction, *container.begin()))>;
```

Comment:
can this one just call the one below with std::map as the first template parameter?

Reply:
Inded, good catch